### ISSUE TITLE
Full screen at startup problematic under Windows

### ISSUE THREAD

**Issue by benfrancis** [mozilla/chromeless#70]
I'm aware this is probably not going to affect many users of Chromeless, but I do think it's a bug.

If your app toggles full screen mode during startup (on document ready) rather than with a button press, it doesn't seem to quite manage to fill the screen on Windows. It works fine if the windows taskbar is already hidden but otherwise the app disappears off the screen slightly and seems to be offset by the approximate height of the Windows taskbar.

Tested on Windows XP & Windows Vista.

**Comment by taboca** [comment]
Input to this thread: 

Tried the solution:

var xulWin = window.QueryInterface(Ci.nsIInterfaceRequestor)
   .getInterface(Ci.nsIWebNavigation).QueryInterface(Ci.nsIDocShellTreeItem)
   .treeOwner.QueryInterface(Ci.nsIInterfaceRequestor)
   .getInterface(Ci.nsIXULWindow);
xulWin.zLevel = xulWin.raisedZ;

And did not work. Ref:  http://forums.mozillazine.org/viewtopic.php?p=2810517

So, this seems not to be related to the always on top issue. Now the interesting thing is that Firefox works, via F11, and I do see Firefox showing sort of the same problem for a while, after the full screen starts and before a secondary step which is when the toolbar collapses. I wonder what you think. 

The browser.js, fullScreen specific, Firefox code is here: 

http://mxr.mozilla.org/mozilla-central/source/browser/base/content/browser.js#3733

**Comment by taboca** [comment]
All right, I think I got a workaround for you, inspired after i looked Firefox behaviour -- my previous comment. 

Just decided to wait ( setTimeout ) a while and make window.innerHeight = window.innerHeight. It worked in a fresh Windows 7 which I just installed. It kinda looks like what Firefox does too -- it goes wrong full screen for half sec and then it fixes itself when the toolbar is collapsed. I will do a seach for possible bug in FF core -- another story. 

Could you test the bug workaround in Windows XP and VIsta?

**Comment by benfrancis** [comment]
Thanks for this. I just briefly tried to test this on Windows XP and unless I'm missing something it doesn't seem to work. Would you expect tests/fullscreen to go full screen in XP because it doesn't for me - I still see the taskbar.

Also, I tried adding "window.innerHeight = window.innerHeight;" to the clock function in my app (which runs every second) and I didn't notice any effect.

Am I missing something or can someone else confirm this doesn't work on XP (and potentially Vista though I haven't had chance to try that yet).

**Comment by taboca** [comment]
Got a machine at Mozilla HQ mountain view to test Windows XP. Indeed. the Solution to Windows 7 does not apply here.

**Comment by taboca** [comment]
The workaround that worked in XP is to simply to call another function after a setTimeout, then in this second function you do the fullscreen. 

```
function test() { 
setTimeout("test2()",1);
}
function test2() { 
   const fullscreen = require("fullscreen");
   fullscreen.toggle(window);
}
```

Please test. Also noticed that Firefox in XP does not go in Full screen with my taskbar locked. And indeed it does with taskbar unlocked and when F11 was pressed after firefox was already launched.

**Comment by benfrancis** [comment]
I suspect it isn't anything to do with having two separate functions, it's just the amount of time which has passed by the time you call the full screen toggle function.

Testing this on XP (in a virtual machine which is probably quite slow) I can get my app to properly go full screen if I set a delay of 2000ms before calling my fullscreen function, but not 1000ms.

Unfortunately this suggests that the amount of time you should wait depends on the system you're running on and so without setting a very long delay it's impossible to know it will work for everyone. It also means the app will appear to start up slowly because of this delay.

So in conclusion this is a workaround, but it isn't a very nice one! I think this will do for now but hopefully someone will be able to figure out the underlying problem.

Thanks for your help taboca

**Comment by taboca** [comment]
I will try to add the full screen to Firefox, in startup time, with no delays, to make sure the problem is over platform or not. Maybe other thing for us to alleviate the problem ( if and in case this is over platform ) is to try to launch the first window in full screen with an appinfo.js attribute.

**Comment by davidmurdoch** [comment]
"try to launch the first window in full screen with an appinfo.js attribute." <- this for sure!

**Comment by benfrancis** [comment]
Webian Shell has a 2 second wait at startup before going full screen to let Windows catch up, but users are still reporting the bug http://getsatisfaction.com/webian/topics/screen_resolution_problem-1d7ltf

I don't really want to extend the delay any further as it creates a perception of a slow startup. Was there any work on an appinfo.js attribute or is it worth re-opening this issue?

**Comment by taboca** [comment]
I think we need to do the test in Firefox. I say : an extension that does the full as soon as possible, to check if we share this with the main product and their views. Let's catch up about this test.

**Comment by benfrancis** [comment]
That sounds like a good idea. I don't have any experience of extension development in Firefox, but I will look into this if I find some time. It would be great to establish exactly where the problem lies as quite a few users are reporting this problem in Windows.
